### ISSUE TITLE
Order of manual mkmk and automatic mark

### ISSUE THREAD

**Issue by simoncozens** [googlefonts/ufo2ft#506]
This is similar to #504 but different. Noto Loopless Thai has a manual `mkmk` feature but a "fully automatic" (i.e. not in the Glyphs source at all) `mark` feature. ufo2ft generates the `mark` code and, because there is no insert point, places it at the end of the feature file. It therefore gets compiled and processed *after* the `mkmk` code, and everything is broken.

Since there is (except perhaps in some truly pathological cases) a hard dependency on `mkmk` being processed after `mark`, we should do the right thing in the default case and ensure that automatic `mark` code gets inserted before any `mkmk` code in the feature file.

**Comment by anthrotype** [comment]
yeah, I guess it'd make more sense to insert mark before this hand-written mkmk instead of appending at the end when no explicit insertion marker is present.

**Comment by simoncozens** [comment]
I think that #512 only partially fixed this bug. I'm still seeing problems with a manual `mkmk` and fully automatic `mark`. Investigating.

**Comment by khaledhosny** [comment]
FWIW, I fix these ordering issues by having an empty`mark`(or `mkmk`) feature that only contains the insert automatic code comment.

**Comment by simoncozens** [comment]
Yeah, I do that too but it's a workaround.

**Comment by anthrotype** [comment]
> I think that https://github.com/googlefonts/ufo2ft/pull/512 only partially fixed this bug... Investigating

@simoncozens is this still an issue?

**Comment by simoncozens** [comment]
Yes, it's causing googlefonts/noto-fonts#2321

**Comment by simoncozens** [comment]
...and when I tried to fix it by adding an empty `mark` feature, I got

```
fontmake: Error: In 'src/NotoSansLepcha.glyphs' -> 'master_ufo/NotoSansLepcha-Regular.designspace' -> 'master_ufo/NotoSansLepcha-Regular.ufo': Compiling UFO failed: <features>:95:35: Unknown markClass @MC_Anchor10
```

This is because Lepcha uses abvm/blwm, and ufo2ft had generated the following erroneous code:

```
feature abvm {
    lookup abvm_mark2base {
        pos base uni1C23
            <anchor 305 583> mark @MC_Anchor10;
...
} abvm;

markClass uni1C2D_1C36 <anchor 0 583> @MC_Anchor10;
...
feature mark {
    lookup mark2base {
        pos base uni25CC
            <anchor 253 583> mark @MC_Anchor10
            <anchor 253 0> mark @MC_Anchor12;
    } mark2base;

} mark;
```

**Comment by anthrotype** [comment]
I'm closing tabs -- is this still open?

**Comment by simoncozens** [comment]
Yep.

**Comment by anthrotype** [comment]
> Yes, it's causing googlefonts/noto-fonts#2321

@simoncozens mind updating this url, it points nowhere -- thanks

**Comment by simoncozens** [comment]
This is now causing Noto Sans Devanagari (plus Latin subset) to fail to build:

```
feature abvm {
# Automatic Code Start

} abvm;

feature blwm {
# Automatic Code Start

} blwm;
```

is turning into:

```
} vatu;

feature abvm {
    lookup abvm_mark2base {
        pos base udeva
            <anchor 299 0> mark @MC_Anchor8;
...

markClass nuktadeva <anchor -283 0> @MC_Anchor8;
...
```

**Comment by simoncozens** [comment]
Swapping the order of abvm/blwm in the feature file somehow managed to make things worse:

```
feature blwm {
    # Automatic Code Start
} blwm;

feature abvm {
    lookup abvm_mark2base {

...
} abvm;

markClass binducandradeva <anchor -130 622> @MC_Anchor2;
markClass candrabindudeva <anchor -130 622> @MC_Anchor2;
markClass anusvaradeva <anchor -126 622> @MC_Anchor2;
```

**Comment by cmyr** [comment]
So I'm coming across this issue trying to understand why `mark` is getting inserted above `mkmk` in the situation where there is an `# Automatic Code` comment for `mkmk` but not for `mark`.

Two questions:

- is the 'dependency' between the two features only about the fact that they are sharing class definitions? Is there anything else going on?
- is `mark` and `mkmk` the only such dependency relationship, or are there others I should consider?
